File: Wayfork/WayforkCore/Sources/WayforkDaemonCore/AtomicFile.swift
import Darwin
import Foundation

/// Root-only files written via temp file + `rename`, so a watcher (sing-box's rule-set
/// reload) and a crash mid-write both see either the old or the new contents.
public enum AtomicFile {
    public enum Error: Swift.Error {
        case openFailed(path: String, errno: Int32)
        case writeFailed(path: String, errno: Int32)
        case renameFailed(path: String, errno: Int32)
    }

    public static func write(_ contents: String, to path: String, mode: mode_t = 0o600) throws {
        try write(Data(contents.utf8), to: path, mode: mode)
    }

    public static func write(_ data: Data, to path: String, mode: mode_t = 0o600) throws {
        let temporary = "\(path).tmp-\(getpid())"
        let descriptor = open(temporary, O_WRONLY | O_CREAT | O_TRUNC | O_CLOEXEC, mode)
        guard descriptor >= 0 else { throw Error.openFailed(path: temporary, errno: errno) }
        var closed = false
        defer {
            if !closed { close(descriptor) }
        }
        _ = fchmod(descriptor, mode)
        var offset = 0
        while offset < data.count {
            let written = data.withUnsafeBytes { buffer in
                Darwin.write(
                    descriptor, buffer.baseAddress!.advanced(by: offset), data.count - offset)
            }
            if written > 0 {
                offset += written
            } else if written == -1 && errno == EINTR {
                continue
            } else {
                let code = errno
                unlink(temporary)
                throw Error.writeFailed(path: temporary, errno: code)
            }
        }
        _ = fsync(descriptor)
        close(descriptor)
        closed = true
        guard rename(temporary, path) == 0 else {
            let code = errno
            unlink(temporary)
            throw Error.renameFailed(path: path, errno: code)
        }
    }
}

File: WayforkWindows/service/internal/core/atomicfile.go
package core

import (
	"fmt"
	"os"
)

// WriteFileAtomic writes `data` to a temporary file next to `path`, syncs it and renames
// it over `path`, so a reader (sing-box's rule-set watcher) and a crash mid-write both see
// either the old or the new contents (docs/design/05-daemon.md, "Files written by the
// daemon"). os.Rename replaces an existing file on Windows too.
func WriteFileAtomic(path string, data []byte, perm os.FileMode) error {
	temporary := fmt.Sprintf("%s.tmp-%d", path, os.Getpid())
	file, err := os.OpenFile(temporary, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, perm)
	if err != nil {
		return fmt.Errorf("atomic write: %w", err)
	}
	if _, err := file.Write(data); err != nil {
		file.Close()
		os.Remove(temporary)
		return fmt.Errorf("atomic write: %w", err)
	}
	if err := file.Sync(); err != nil {
		file.Close()
		os.Remove(temporary)
		return fmt.Errorf("atomic write: %w", err)
	}
	if err := file.Close(); err != nil {
		os.Remove(temporary)
		return fmt.Errorf("atomic write: %w", err)
	}
	if err := os.Rename(temporary, path); err != nil {
		os.Remove(temporary)
		return fmt.Errorf("atomic write: %w", err)
	}
	return nil
}

// WriteStringAtomic is WriteFileAtomic for text.
func WriteStringAtomic(path, contents string, perm os.FileMode) error {
	return WriteFileAtomic(path, []byte(contents), perm)
}
